File: guidedFilter.py
#-*-coding: utf-8-*-
##author: wxs  Wuhan University
##data: 2021-3-5
##Implementation of Guided Filter by KaiMing He

import numpy as np
import cv2
from matplotlib import pyplot as plt

##利用box filter来快速地计算每个窗口内的像素总和
##Box Filter通过积分图的思想将这一过程优化，从O(MN)变成O(4)
##输入：引导图像和窗口的半径
def boxFilter(imgSrc, r):
    img_h, img_w = imgSrc.shape
    imgDst = np.zeros([img_h, img_w], np.float32)

    ##首先累加y方向(↓)
    imgCum = np.cumsum(imgSrc, 0)
    ##注意:通过不同位置的累加值的差相减来得到每个像素在当前窗口内，y方向上的窗口累加值
    imgDst[0:r + 1, :] = imgCum[r:2*r + 1, :]
    imgDst[r + 1:img_h-r, :] = imgCum[2*r + 1:img_h, :] - imgCum[0:img_h-2*r - 1, :]
    imgDst[img_h - r:img_h, :] = np.tile(imgCum[img_h-1, :].reshape([1, -1]), [r, 1]) - imgCum[img_h-2*r - 1:img_h-r - 1, :]

    ##然后在累加完y方向的基础上再按x方向(→)累加，来实现整个窗口二维的累加
    imgCum = np.cumsum(imgDst, 1)
    ##同样通过x方向不同位置的累加值的差来计算窗口内所有像素的和
    imgDst[:, 0:r+1] = imgCum[:, r:2*r+1]
    imgDst[:, r+1:img_w-r] = imgCum[:, 2*r+1:img_w] - imgCum[:, 0:img_w-2*r-1]
    imgDst[:, img_w-r:] = np.tile(imgCum[:, img_w-1].reshape([-1, 1]), [1, r]) - imgCum[:, img_w-2*r-1:img_w-r-1]

    return imgDst

##输入：待滤波图像、引导图像、输出图像、窗口半径、
def guidedFilter(I, p, r, eps):
    img_h, img_w = p.shape
    ##N矩阵的每个像素计算了当前像素作为窗口中心像素时窗口内像素的个数
    N = boxFilter(np.ones_like(I), r)

    mean_I = boxFilter(I, r) / N
    float_I = np.float32(I)
    mean_II = boxFilter(float_I**2, r) / N
    ##mean_II = boxFilter(I*I, r) / N   ##这里I如果是uint8的,I*I的结果会溢出255,所以转为float32处理
    var_I = mean_II - mean_I**2
    
    mean_p = boxFilter(p, r) / N
    float_Ip = float_I * np.float32(p)
    mean_Ip = boxFilter(float_Ip, r) / N
    cov_Ip = mean_Ip - mean_I * mean_p
    ##这里del是尽快释放一些空间，图像不大的话可以不用
    del mean_Ip, mean_II

    a = cov_Ip / (var_I + eps)
    del cov_Ip, var_I
    b = mean_p - a * mean_I
    del mean_p, mean_I

    mean_a = boxFilter(a, r) / N
    mean_b = boxFilter(b, r) / N

    outputImg = mean_a * I + mean_b
    return outputImg

 

File: guidedFilter.cpp
#include "guidedFilter.h"

//将矩阵按照某一个方向逐个累加，axis==0 沿着y轴累加(↓)，axis==1沿着x轴累加(→)
template<class T>
Mat accumulateSum(Mat &srcImg, int32 axis = 0)
{
	int32 img_h = srcImg.rows, img_w = srcImg.cols;
	Mat cumImg = Mat::zeros(img_h, img_w, CV_32FC1);

	//从上而下累加
	if (axis == 0)
	{
		//第0行
	//#pragma omp parallel for
	//	for (int32 j = 0; j < img_w; ++j)
	//	{
	//		cumImg.at<float32>(0, j) = srcImg.at<T>(0, j);
	//	}
		srcImg.row(0).copyTo(cumImg.row(0));

		//第1行到最后一行
		for (int32 i = 1; i < img_h; ++i)
		{
	#pragma omp parallel for
			for (int32 j = 0; j < img_w; ++j)
			{
				cumImg.at<float32>(i, j) = cumImg.at<float32>(i - 1, j) + srcImg.at<T>(i, j);
			}
		}
	}
	////从左到右累加
	else
	{
	//#pragma omp parallel for
	//	for (int32 i = 0; i < img_h; ++i)
	//	{
	//		cumImg.at<float32>(i, 0) = srcImg.at<T>(i, 0);
	//	}
		srcImg.col(0).copyTo(cumImg.col(0));

		for (int32 j = 1; j < img_w; ++j)
		{
	#pragma omp parallel for
			for (int32 i = 0; i < img_h; ++i)
			{
				cumImg.at<float32>(i, j) = cumImg.at<float32>(i, j - 1) + srcImg.at<T>(i, j);
			}
		}
	}
	return cumImg;
}

Mat matAccumulateSum(Mat &srcImg, int32 axis = 0)
{
	int32 dType = srcImg.depth();
	if (dType == CV_8S)
	{
		return accumulateSum<int8>(srcImg, axis);
	}
	else if (dType == CV_16U)
	{
		return accumulateSum<uint16>(srcImg, axis);
	}
	else if (dType == CV_16S)
	{
		return accumulateSum<int16>(srcImg, axis);
	}
	else if (dType == CV_32S)
	{
		return accumulateSum<int32>(srcImg, axis);
	}
	else if (dType == CV_32F)
	{
		return accumulateSum<float32>(srcImg, axis);
	}
	else
	{
		return accumulateSum<uint8>(srcImg, axis);
	}
}

Mat boxFilter(Mat &srcImg, int32 r)
{
	int32 img_h = srcImg.rows, img_w = srcImg.cols;
	Mat dstImg = Mat::zeros(img_h, img_w, CV_32F);
	Mat cumImg = matAccumulateSum(srcImg, 0);

	//按y方向累加值计算窗口元素y方向上的和
	cumImg.rowRange(r, 2 * r + 1).copyTo(dstImg.rowRange(0, r + 1));
	Mat res = cumImg.rowRange(2 * r + 1, img_h) - cumImg.rowRange(0, img_h - 2 * r - 1);
	res.copyTo(dstImg.rowRange(r + 1, img_h - r));
	for (int32 k1 = img_h - r, k2 = img_h - 2 * r - 1; k1 < img_h && k2 < img_h - r - 1; ++k1, ++k2)
	{
		Mat tmp = cumImg.rowRange(img_h - 1, img_h) - cumImg.rowRange(k2, k2 + 1);
		tmp.copyTo(dstImg.rowRange(k1, k1 + 1));
	}

	//按x方向累加值计算窗口元素x方向上的和
	cumImg = matAccumulateSum(dstImg, 1);
	cumImg.colRange(r, 2 * r + 1).copyTo(dstImg.colRange(0, r + 1));
	res = cumImg.colRange(2 * r + 1, img_w) - cumImg.colRange(0, img_w - 2 * r - 1);
	res.copyTo(dstImg.colRange(r + 1, img_w - r));
	for (int32 k1 = img_w - r, k2 = img_w - 2 * r - 1; k1 < img_w && k2 < img_w - r - 1; ++k1, ++k2)
	{
		Mat tmp = cumImg.colRange(img_w - 1, img_w) - cumImg.colRange(k2, k2 + 1);
		tmp.copyTo(dstImg.colRange(k1, k1 + 1));
	}

	return dstImg;
}

Mat guidedFilter(Mat &I, Mat &p, int32 r, float32 eps)
{
	int32 img_h = I.rows, img_w = I.cols;
	Mat oneMat = Mat::ones(img_h, img_w, CV_8UC1);
	Mat N = boxFilter(oneMat, r);
	
	Mat mean_I = boxFilter(I, r) / N;
	Mat float_I;
	I.convertTo(float_I, CV_32FC1);
	Mat II = float_I.mul(float_I);
	Mat mean_II = boxFilter(II, r) / N;
	Mat var_I = mean_II - mean_I.mul(mean_I);
	II.release();
	mean_II.release();
	
	Mat mean_p = boxFilter(p, r) / N;
	Mat float_p;
	p.convertTo(float_p, CV_32FC1);
	Mat Ip = float_I.mul(float_p);
	float_p.release();
	Mat mean_Ip = boxFilter(Ip, r) / N;
	Ip.release();
	Mat cov_Ip = mean_Ip - mean_I.mul(mean_p);
	mean_Ip.release();

	Mat a = cov_Ip / (var_I + eps);
	Mat b = mean_p - a.mul(mean_I);
	cov_Ip.release();
	var_I.release();
	mean_p.release();
	mean_I.release();

	Mat mean_a = boxFilter(a, r) / N;
	Mat mean_b = boxFilter(b, r) / N;

	a.release();
	b.release();

	Mat outputImg = mean_a.mul(float_I) + mean_b;
	return outputImg;
}

